src/server/game/Behavior/ActionScript/ActionThreadStep.cpp

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ActionThreadStep.h"
+#include "ActionScript.h"
+
+class NullStep : public ActionThreadStep
+{
+    public:
+        NullStep(ActionThread const& thread, ActionScriptStep const& temp) : ActionThreadStep(thread, temp) {}
+        void Initialize() override {}
+        Optional<bool> Evaluate() override {
+            return {};
+        }
+        void Abort() override {}
+};
+
+/*static*/ std::unique_ptr<ActionThreadStep> ActionThreadStep::StepTo(ActionThread const& thread, ActionScriptStep const& stepTemplate)
+{
+    ActionThreadStep* step;
+    switch (stepTemplate.StepType)
+    {
+        case ACTIONSCRIPT_NULL:
+            step = new NullStep(thread, stepTemplate);
```

Comment:
If "steps" are anything like SAI's actions or even events, dynamic allocations every time a script is going to execute an action (or event) seems very detrimental to performance. Consider using std::variant, so no allocations will need to be done and all possible steps can live within the same already-allocated memory. Although it will probably make it very inconvenient to declare said variant, since you would need to list all possible step implementations in it. Another way would be to have a pre-allocated memory arena that will host all steps (with "placement new"s), however that would require defining strict limitations on step sizes (nothing a static_assert can't handle) to make sure they can even fit inside the arena.

Reply:
Good observations. Variant would be a complete pain to work with.

Looking into placement new into a pre-allocated region in the `ActionThread`.